### PR TITLE
feature-xray-and-bug-fixes

### PR DESCRIPTION
**Pull request by lavie**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -13,29 +13,38 @@ DOCKERARGS := -e AWS_ACCESS_KEY_ID=$(AWS_ACCESS_KEY_ID) \
...
 	$(SLS) invoke $(INVOKE_ARGS) -f $* $(data) -s $(stage) -p input.json
 
-invoke-find-unavailable-instance-types: data='-d { "region": "$(region)" }'
-
 .PHONY: bash
 bash: build
 	$(DOCKER) run -it --rm $(DOCKERARGS) --entrypoint /bin/bash $(IMAGE)
 
-update-functions: deploy-find-unavailable-instance-types deploy-patch-asg
+update-functions: $(foreach func, $(FUNCTIONS), deploy-$(func))
```
Comment (bergundy) [inline Makefile]: .PHONY: update-functions lint-fix ...

File: get_lc_instance_type.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+monkey_patch_botocore_for_xray()
+
+cache = dict()
+
+
+def get_lc_instance_type(client, lc_name):
+    desc = client.describe_launch_configurations(
+        LaunchConfigurationNames=[lc_name]
+    )
+    if not len(desc['LaunchConfigurations']):
```
Comment (bergundy) [inline get_lc_instance_type.py]: Could `LaunchConfigurations` be missing in `desc`? If so it'll raise a `KeyError`

File: get_lc_instance_type.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        raise Exception('LC %s not found' % lc_name)
+    lc = desc['LaunchConfigurations'][0]
+    used_instance_type = lc['InstanceType']
+    print("%s uses instance type %s" % (lc_name, used_instance_type))
```
Comment (bergundy) [inline get_lc_instance_type.py]: If you're using python3.6 it is recommended to use the new string formatting syntax: f'{lc_name} uses instance type {used_instance_type}'

File: get_lc_instance_type.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+from utils import http_error_handling
+import boto3
+from fleece.xray import monkey_patch_botocore_for_xray
+import os
+
+monkey_patch_botocore_for_xray()
+
+cache = dict()
```
Comment (bergundy) [inline get_lc_instance_type.py]: You could simply use `functools.lru_cache`

File: utils.py [hunk 5]
```diff
@@ -18,13 +20,20 @@ def func_wrapper(*args, **kwds):
...
-    client.invoke(FunctionName="asgard-%s-%s" % (stage, func),
-                  InvocationType="Event",
-                  Payload=bytes(j, "utf8")
-                  )
-
+    res = client.invoke(FunctionName="asgard-%s-%s" % (stage, func),
+                        InvocationType="RequestResponse" if block else "Event",
+                        Payload=bytes(j, "utf8")
+                        )
+    if block:
```
Comment (bergundy) [inline utils.py]: What is this `block` argument and why do we need it? (please document)

File: get_lc_instance_type.py (hunk 2 above)
Comment (lavie) [reply get_lc_instance_type.py]: always returns an array, even if empty:
```
saws> aws autoscaling describe-launch-configurations --region us-east-2
{
    "LaunchConfigurations": []
}

```

File: utils.py (hunk 5 above)
Comment (lavie) [reply utils.py]: as you can see in the live above, it determines whether the invocation type is blocking or not. if so, I return the result.

**Pull request merged**